File: backend/app/schemas/vision.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


VisionTaskStatus = Literal["PROCESSING", "DONE", "FAILED"]


class VisionDetectionBox(BaseModel):
    x1: float
    y1: float
    x2: float
    y2: float


class VisionDetectionItem(BaseModel):
    label: str
    confidence: float
    bbox: VisionDetectionBox | None = None


class VisionTaskResponse(BaseModel):
    taskId: str
    status: VisionTaskStatus
    source: str
    imageUrl: str
    diseaseType: str | None = None
    confidence: float | None = None
    detections: list[VisionDetectionItem] = Field(default_factory=list)
    engine: str | None = None
    device: str | None = None
    fallbackOccurred: bool | None = None
    error: str | None = None
    queuedAt: datetime | None = None
    processedAt: datetime | None = None
    createdAt: datetime
    updatedAt: datetime


class VisionTaskListResponse(BaseModel):
    items: list[VisionTaskResponse]


class VisionRuntimeResponse(BaseModel):
    mode: str
    engine: str
    preferredDevice: str
    activeDevice: str
    fallbackOccurred: bool
    storageBackend: str
    queueKey: str
    queueDepth: int
    maxUploadMb: int

File: frontend/src/types/vision.ts
export type VisionTaskStatus = "PROCESSING" | "DONE" | "FAILED";

export type VisionDetectionBox = {
  x1: number;
  y1: number;
  x2: number;
  y2: number;
};

export type VisionDetectionItem = {
  label: string;
  confidence: number;
  bbox?: VisionDetectionBox | null;
};

export type VisionTask = {
  taskId: string;
  status: VisionTaskStatus;
  source: string;
  imageUrl: string;
  diseaseType?: string | null;
  confidence?: number | null;
  detections: VisionDetectionItem[];
  engine?: string | null;
  device?: string | null;
  fallbackOccurred?: boolean | null;
  error?: string | null;
  queuedAt?: string | null;
  processedAt?: string | null;
  createdAt: string;
  updatedAt: string;
};

export type VisionTaskListPayload = {
  items: VisionTask[];
};

export type VisionRuntimePayload = {
  mode: string;
  engine: string;
  preferredDevice: string;
  activeDevice: string;
  fallbackOccurred: boolean;
  storageBackend: string;
  queueKey: string;
  queueDepth: number;
  maxUploadMb: number;
};

export type VisionWsUrlPayload = {
  url: string;
};

export type VisionTaskEventPayload = {
  type: "vision.connected" | "vision.pong" | "vision.task.accepted" | "vision.task.updated";
  task?: VisionTask;
  queueDepth?: number;
  connectedAt?: string;
  runtime?: VisionRuntimePayload;
  at?: string;
};
